### PR TITLE
Support for if, else, then rules

### PR DESCRIPTION
**Pull request by Juansasa**
Added if then else logic to resolve schemas

### Reasons for making this change

If else, then support is crucial for doing dynamic form while avoiding lots of complicated logics. Forms will be much easier to generate as well. 

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [x] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by Juansasa** [comment]
Inspired by #2466 but without other unrelated fixes

**Comment by levyitay** [comment]
Looks good please merge

File: packages/core/src/utils.js [hunk 1]
```diff
@@ -662,12 +662,37 @@ export function stubExistingAdditionalProperties(
   return schema;
 }
 
+const _resolveCondition = (schema, rootSchema, formdata) => {
```
Comment (levyitay) [inline packages/core/src/utils.js:665]: none of the private method has _ in the method name, maybe best to keep it without

**Review by levyitay — CHANGES_REQUESTED** [review]
Looks good to me, only 1 minor change in the method's name

File: packages/core/src/utils.js (hunk 1 above)
Comment (Juansasa) [reply packages/core/src/utils.js]: Done

**Review by levyitay — APPROVED** [review]

**Comment by nickgros** [comment]
I don't know if I would say the other changes in #2466 are "unrelated". At first glance, this implementation can't handle `$ref`s in if/then/else, e.g. [this test](https://github.com/nickgros/react-jsonschema-form/blob/b6c53c89fd794cad9642d47232ff3aa5be4635ef/packages/core/test/ifthenelse_test.js#L117).

**Comment by Juansasa** [comment]
> I don't know if I would say the other changes in #2466 are "unrelated". At first glance, this implementation can't handle `$ref`s in if/then/else, e.g. [this test](https://github.com/nickgros/react-jsonschema-form/blob/b6c53c89fd794cad9642d47232ff3aa5be4635ef/packages/core/test/ifthenelse_test.js#L117).

I'd say optimizations and other 

> I don't know if I would say the other changes in #2466 are "unrelated". At first glance, this implementation can't handle `$ref`s in if/then/else, e.g. [this test](https://github.com/nickgros/react-jsonschema-form/blob/b6c53c89fd794cad9642d47232ff3aa5be4635ef/packages/core/test/ifthenelse_test.js#L117).

You are correct abut the $ref issue, new test added.
What i mean with unrelated is the other refactoring/improvements to validation which i think is non-essential for resolving ifThen schemas and should belong in a separate PR. 

But i might be wrong. My intention is to make a minimal change set to get this approve as quickly as possible as we are already using this on a separate project. Sorry if i sidestep you on this

**Comment by nickgros** [comment]
> What i mean with unrelated is the other refactoring/improvements to validation which i think is non-essential for resolving ifThen schemas and should belong in a separate PR.

I totally get it-- #2466 inherited a lot of cruft from #1666 (for instance, that #1666 dropped `json-schema-merge-allof` in favor of a custom solution, but it looks like you're using it with no problems). 

My use cases anticipate complex schemas, which is why I didn't decouple optimizations from that PR. I think some form of optimization will still be necessary if this PR, and not #2466, is merged.

**Comment by epicfaace** [comment]
Thanks for both your PRs @Juansasa and @nickgros ! I don't really have the bandwidth to compare implementations at the moment, would someone be able to carefully review both PRs and recommend the best plan of action (i.e., which one to merge, or if both PRs should be combined in some way)?

Ultimately, I'd like to have a PR merged that has the most extensive if-else-then support, and probably we want to do that in the simplest way with the least changes.

**Comment by AlimovSV** [comment]
Any updates? We are looking forward to `if-then-else` implementation will be released.

**Comment by jmsaulnier** [comment]
> Thanks for both your PRs @Juansasa and @nickgros ! I don't really have the bandwidth to compare implementations at the moment, would someone be able to carefully review both PRs and recommend the best plan of action (i.e., which one to merge, or if both PRs should be combined in some way)?
> 
> Ultimately, I'd like to have a PR merged that has the most extensive if-else-then support, and probably we want to do that in the simplest way with the least changes.

any updates? Thanks 🙏

**Comment by heath-freenome** [comment]
> Thanks for both your PRs @Juansasa and @nickgros ! I don't really have the bandwidth to compare implementations at the moment, would someone be able to carefully review both PRs and recommend the best plan of action (i.e., which one to merge, or if both PRs should be combined in some way)?
> 
> Ultimately, I'd like to have a PR merged that has the most extensive if-else-then support, and probably we want to do that in the simplest way with the least changes.

@epicfaace why not merge the simpler implementation then @nickgros can rebase his changes on top and deal with the corner cases?

File: packages/core/src/utils.js [hunk 2]
```diff
@@ -662,12 +662,40 @@ export function stubExistingAdditionalProperties(
   return schema;
 }
 
+const resolveCondition = (schema, rootSchema, formdata) => {
```
Comment (epicfaace) [inline packages/core/src/utils.js:665]: Rename formdata to `formData`

Comment (epicfaace) [reply packages/core/src/utils.js:665]: Add a comment describing what this function does

**Comment by epicfaace** [comment]
> @epicfaace why not merge the simpler implementation

Good idea! I took a quick pass, I could use some help reviewing

**Comment by epicfaace** [comment]
@nickgros any problems with that approach? is it possible that this PR may not be backwards compatible with https://github.com/rjsf-team/react-jsonschema-form/pull/2466?

**Comment by nickgros** [comment]
> @nickgros any problems with that approach? is it possible that this PR may not be backwards compatible with #2466?

This is fine with me. Many of the changes I made in #2466 are performance optimizations that I needed for my use case (I need to support very large/complex schemas with conditions). The changes I made may not necessarily even be the right approach to solve those problems.

Since this PR is more straightforward and seems to enable this (very valuable!) feature, I think it's worth moving forward for the sake of improving this library. Optimizations can always come later.

File: packages/core/test/utils_test.js [hunk 3]
```diff
@@ -2388,6 +2388,210 @@ describe("utils", () => {
         });
       });
     });
+
+    describe("ifElseThen", () => {
```
Comment (nickgros) [inline packages/core/test/utils_test.js:2392]: "ifElseThen" -> "ifThenElse"

File: packages/playground/src/samples/index.js [hunk 4]
```diff
@@ -52,6 +53,7 @@ export const samples = {
   "Any Of": anyOf,
   "One Of": oneOf,
   "All Of": allOf,
+  "If Else Then": ifElseThen,
```
Comment (nickgros) [inline packages/playground/src/samples/index.js:56]: "If Else Then" -> "If Then Else"

**Comment by jmsaulnier** [comment]
Any updates on merging this PR? @Juansasa @nickgros @epicfaace 
Thanks!

**Comment by ksbrar** [comment]
Also eagerly awaiting this feature!

**Comment by heath-freenome** [comment]
@Juansasa consider joining the rjsf weekly meeting on 2/11 to discuss your PR

**Comment by epicfaace** [comment]
Great. Please add to the agenda!

Ashwin Ramaswami

**Comment by Juansasa** [comment]
Sorry guys missed the notifications.
I fixed the tests and a bug based on our internal usage.
Let me know if you guys need further assistance

File: packages/core/src/utils.js [hunk 5]
```diff
@@ -665,12 +665,40 @@ export function stubExistingAdditionalProperties(
...
+
+  const conditionalSchema = isValid(expression, formdata, rootSchema)
+    ? then
+    : otherwise;
+
+  if (conditionalSchema) {
+    return retrieveSchema(
+      mergeSchemas(
+        resolvedSchemaLessConditional,
+        retrieveSchema(conditionalSchema, rootSchema, formdata)
```
Comment (nickgros) [inline packages/core/src/utils.js:684]: Noticed that this was swapped (presumably to fix a bug?). Is there a test case we could add that would catch a regression?

File: packages/core/test/utils_test.js [hunk 6]
```diff
@@ -2403,6 +2403,210 @@ describe("utils", () => {
...
+          properties: {
+            animal: {
+              enum: ["Cat", "Fish"],
+            },
+            food: { type: "string", enum: ["meat", "grass", "fish"] },
+          },
+          required: ["animal", "food"],
+        });
+      });
+      it.only("should resolve $ref", () => {
```
Comment (nickgros) [inline packages/core/test/utils_test.js:2541]: remove .only

File: packages/core/src/utils.js (hunk 5 above)
Comment (epicfaace) [reply packages/core/src/utils.js:684]:
```
{
if: ...,
then: {minLength: 5},
minLength: 10
}
```

Expected behavior: inner property (`minLength: 5`) should override outer property (`minLength: 10`). We should add a test case for this

**Comment by epicfaace** [comment]
Thanks for your help @Juansasa !!! 🎉  @nickgros will take a couple of tiny changes forward in https://github.com/rjsf-team/react-jsonschema-form/pull/2700 and we will make sure you get credit when the PR is merged.

**Pull request closed without merging**